pyvrp/cpp/search/LocalSearch.cpp

```diff
@@ -251,47 +254,110 @@ void LocalSearch::applyEmptyRouteMoves(Route::Node *U,
 void LocalSearch::applyOptionalClientMoves(Route::Node *U,
                                            CostEvaluator const &costEvaluator)
 {
-    auto const uClient = U->client();
-    ProblemData::Client const &uData = data.location(uClient);
+    ProblemData::Client const &uData = data.location(U->client());
+
+    auto mustInsert = uData.required;
+    if (uData.group)
+    {
+        auto const &group = data.group(uData.group.value());
+        auto const pred = [&](auto client) { return nodes[client].route(); };
+        auto const groupInSol = std::any_of(group.begin(), group.end(), pred);
+
+        if (groupInSol)  // then we'll have to wait for the group moves to
+            return;      // determine what to do.
+
+        // U must be inserted when it is in a required mutually exclusive client
+        // group that currently is not in the solution at all.
+        assert(group.mutuallyExclusive);
+        mustInsert = group.required && !groupInSol;
+    }
```

Comment:
Just to summarize for myself: The reason why this part is here instead of `GroupMove` is to prevent clients from being inserted when another group client is already in the solution. The original `applyOptionalClientMoves` would just insert the optional client if improving, which is redundant if they are being removed afterwards again.

But the main reason seems to be that this always ensures that all groups are part of the solution before applying group moves.